File: java-api/src/main/java/com/fagerland/javaapi/filters/JwtFilter.java
package com.fagerland.javaapi.filters;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.Jwt;
import io.jsonwebtoken.Jwts;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.GenericFilterBean;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;

public class JwtFilter extends GenericFilterBean {

    public JwtFilter() {

    }

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
        HttpServletRequest req = (HttpServletRequest) servletRequest;
        String bearerToken = req.getHeader("Authorization");
        if (bearerToken != null) {
            String token = bearerToken.substring(7, bearerToken.length());

            Jws<Claims> claims = Jwts.parser().setSigningKey("secret").parseClaimsJws(token);
            if (!claims.getBody().getExpiration().before(new Date())) {

                String username = claims.getBody().getSubject();
                Authentication auth = new UsernamePasswordAuthenticationToken(username, null, new ArrayList<>());
                SecurityContextHolder.getContext().setAuthentication(auth);
            }
        }
        filterChain.doFilter(servletRequest, servletResponse);
    }
}

File: kotlin-maven-api/src/main/kotlin/com/fagerland/brotime/filters/JwtFilter.kt
package com.fagerland.brotime.filters

import com.fagerland.brotime.repositories.UserRepository
import com.fagerland.brotime.services.JwtService
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jws
import io.jsonwebtoken.Jwts
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.GenericFilterBean
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


class JwtFilter(
    private val jwtService: JwtService
) : GenericFilterBean() {
    override fun doFilter(servletRequest: ServletRequest?, servletResponse: ServletResponse?, filterChain: FilterChain?) {
        val req: HttpServletRequest = servletRequest as HttpServletRequest
        val res: HttpServletResponse = servletResponse as HttpServletResponse
        val validUsername = jwtService.getUsernameFromRequest(req)
        if (validUsername != null) {
            val auth: Authentication = UsernamePasswordAuthenticationToken(validUsername, null, arrayListOf())
            SecurityContextHolder.getContext().authentication = auth
            filterChain?.doFilter(servletRequest, servletResponse)
        } else {
            res.sendError(HttpServletResponse.SC_UNAUTHORIZED)
            return
        }
    }
}
